feat: Add `hs doctor` command

joe-yeager opened the pull request:
## Description and Context
<!-- Provide a summary of what has changed -->
<!-- Provide links to relevant discussions or documentation to promote understanding and addressing this PR -->
<!-- Describe any packages you'd like to add and the reasons why. -->
Adds a `hs doctor` command to aid users in self diagnosing problems with their CLI, CLI configuration, and Project configuration

## Screenshots

### Success screenshots
#### Outside a project
![Screenshot 2024-11-07 at 3 36 26 PM](https://github.com/user-attachments/assets/cbd7c46d-a67e-4bea-a5b2-843e6bc7def7)

#### Inside a project 
![Screenshot 2024-11-07 at 3 38 37 PM](https://github.com/user-attachments/assets/de45b5eb-f8d3-4472-8167-5623ccbfc1da)

### Failure screenshots
#### Inactive account

![Screenshot 2024-11-07 at 3 40 57 PM](https://github.com/user-attachments/assets/b5bfab8d-6310-4d3f-9967-d59c54e8ebcd)

#### Missing dependency installs
![Screenshot 2024-11-07 at 3 41 53 PM](https://github.com/user-attachments/assets/efd0d9ca-72af-405a-8b53-b8f3f39bf976)

#### Port 8080 in use
![Screenshot 2024-11-07 at 3 48 02 PM](https://github.com/user-attachments/assets/adb49628-4323-4a02-a18d-343c49d5c261)

bmatto commented:
neat.

joe-yeager commented on .github/workflows/build.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
       - name: Verify package dependencies
         run: |
-          npx depcheck --ignores=semver
+          npx depcheck --ignores="semver,@types/jest"
```
I'm debating adding `@types/*` to the ignore because there are valid reasons to include types without ever importing them

brandenrodgers commented on commands/doctor.ts:
```diff
@@ -0,0 +1,73 @@
+  const doctor = new Doctor();
+
+  try {
+    trackCommandUsage('doctor', undefined, doctor.accountId);
```
The trackCommandUsage command itself should handle the try/catch and debugging. So you should be able to just call it directly without the wrapper.

brandenrodgers commented on commands/doctor.ts:
```diff
@@ -0,0 +1,73 @@
+
+  try {
+    fs.writeFileSync(outputFile, JSON.stringify(output, null, 4));
+    logger.success(`Output written to ${outputFile}`);
```
Looks like some copy made it in here

brandenrodgers commented on lang/en.lyaml:
```diff
@@ -1418,3 +1425,52 @@ en:
+        unableToDetermine: "Unable to determine if the portal is active"
+        pak:
+          invalid: "Personal access key is invalid"
+          valid: "Personal Access Key valid. {{ link }}"
```
should this have "is"?

brandenrodgers commented on lang/en.lyaml:
```diff
@@ -275,6 +275,13 @@ en:
                 update: "Your schema has been updated in account \"{{ accountId }}\""
                 viewAtUrl: "Schema can be viewed at {{ url }}"
+    doctor:
+      describe: "The doctor is in"
```
😆 placeholder?

One suggestion is: "Run diagnostics to detect common CLI issues"

joe-yeager replied:
Yeah, all the copy is awaiting input from UX 😄 

I should have mentioned that

brandenrodgers commented on lang/en.lyaml:
```diff
@@ -1418,3 +1425,52 @@ en:
+        missingDependencies: "missing dependencies in {{#cyan}}{{ dir }}{{/cyan}}"
+        missingDependenciesSecondary: "Run {{ command }} to install all project dependencies locally"
+        unableToDetermine: "Unable to determine if dependencies are installed {{ dir }}"
+        success: "App dependencies are installed an up to date"
```
Typo on "and" here

brandenrodgers commented on lang/en.lyaml:
```diff
@@ -1418,3 +1425,52 @@ en:
+        unableToDetermine: "Unable to determine if dependencies are installed {{ dir }}"
+        success: "App dependencies are installed an up to date"
+      files:
+        invalidJson: "invalid JSON in {{#cyan}}{{ filename }}{{/cyan}}"
```
iirc we don't typically use cyan for filepaths. We use bold. Was this from the mocks?

joe-yeager replied:
I just messaged Sandra to confirm

joe-yeager replied:
Changed them to bold

brandenrodgers commented on lang/en.lyaml:
```diff
@@ -1418,3 +1425,52 @@ en:
+        cliConfig:
+          header: "CLI configuration"
+          configFileSubHeader: "Config File: {{#cyan}}{{ filename }}{{/cyan}}"
+          defaultAccountSubHeader: "Default Account: {{#cyan}}{{ accountName }}[{{ accountType }}]({{ accountId }}) {{/cyan}}"
```
We have a `uiAccountDescription` ui util for logging out accounts

brandenrodgers commented on lang/en.lyaml:
```diff
@@ -1418,3 +1425,52 @@ en:
+          projectDirSubHeader: "Project dir: {{#cyan}}{{ projectDir }}{{/cyan}}"
+          projectNameSubHeader: "Project name: {{#cyan}}{{ projectName }}{{/cyan}}"
+        counts:
+          errors: '{{#bold}}Errors{{/bold}}: {{ count }}'
```
nit, but should the colons be included in the bold here?

brandenrodgers commented on lib/dependencyManagement.ts:
```diff
@@ -150,9 +150,18 @@ async function getProjectPackageJsonLocations() {
   return packageParentDirs;
 }
 
+export async function packagesNeedInstalled(directory) {
```
Wonder if we could come up with a slightly more informative name for this one. Some suggestions are `areDependenciesUpToDate()`, `detectAvailableDependencyUpdates()`, `areNewDependenciesAvailable()`

joe-yeager replied:
I agree we could have a better name, but this isn't checking for updates, it is checking the any packages would be installed.  `update` implies that there may new versions available and that is not a part of the check.  Maybe something like `hasMissingPackages`

brandenrodgers commented on lib/doctor/DiagnosticInfoBuilder.ts:
```diff
@@ -0,0 +1,206 @@
+  diagnosis?: string;
+}
+
+const configFiles = [
```
There are other files that can exist in projects too, like theme.json and cms-assets.json. I wonder though, rather than hardcode this list (which we'd have to continually update as new configs get added) could we instead just search the project for files with `.json` extensions and then run validation against all of them? The assumption being that any json file in the project is worth validating.

joe-yeager replied:
Yeah, in the lack of a BE to provide these dynamically we kind of had to hardcode them 😞 

We could just validate all of the `json` files in the project if we wanted to.  That would mean that any user defined JSON files would also be validated, so it might be a lot in some of the larger projects.  But it scales with the addition of new files better.

joe-yeager replied:
I changed it to validate all of the JSON files because it really doesn't add that much overhead.  I updated this list of have the two files you added so it adds them to the `configFiles` list which is just a QOL item for when users submit the output file

brandenrodgers commented on lib/doctor/DiagnosticInfoBuilder.ts:
```diff
@@ -0,0 +1,206 @@
+      path: mainModule?.path,
+      config: getConfigPath(),
+      versions: {
+        '@hubspot/cli': pkg.version,
```
Not a blocker, but we reference `@hubspot/cli` in enough places that it might be worthwhile to add it as a constant somewhere

brandenrodgers commented on lib/doctor/Doctor.ts:
```diff
@@ -0,0 +1,314 @@
+    try {
+      const readFile = util.promisify(fs.readFile);
+      const fileContents = await readFile(filename);
+      JSON.parse(fileContents.toString());
```
Could this be shortened to `JSON.parse(fs.readFileSync(filePath, 'utf-8'))`?

joe-yeager replied:
That will have performance ramifications.  I strayed away from all the `fs.*Sync` methods intentionally because they are blocking and don't allow for the checks to run async in parallel.

brandenrodgers commented:
I deleted my hubspot.config.yml file and ran the command. It correctly threw an error, but I think we could give a more useful message here. This is what the output looks like:
<img width="375" alt="image" src="https://github.com/user-attachments/assets/662ec19c-6838-42dd-943f-7ae165e04abb">
It might be nice to make sure that the hubspot.config.yml file exists before continuing to the access token checks

joe-yeager commented on lib/doctor/Doctor.ts:
```diff
@@ -0,0 +1,314 @@
+
+  private async checkIfPortsAreAvailable() {
+    const localI18nKey = `${i18nKey}.port`;
+    if (await isPortManagerServerRunning()) {
```
Yeah, I agree that this can be augmented to check if the service running is the actual port manager service.  I think that will be a good candidate for the next iteration!

joe-yeager replied:
Since we had an issue to update the portManager shutdown flow once [this LDL PR is merged](https://github.com/HubSpot/hubspot-local-dev-lib/pull/209), I can update this to check if the port is available, and then check if it is the port manager

joe-yeager commented on lib/doctor/Doctor.ts:
```diff
@@ -0,0 +1,314 @@
+const { i18n } = require('../lang');
+const { uiLink } = require('../ui');
+
+const minMajorNodeVersion = 18;
```
Kind of.  I debated using the `engines.node` prop and stripping the non-numeric values, but `engines.node` supports syntax like `>=18 <=21`.  If we think we won't get in a place where we leverage the more complex syntax, I think reading it from there makes sense

joe-yeager commented on lib/dependencyManagement.ts:
```diff
@@ -150,9 +150,18 @@ async function getProjectPackageJsonLocations() {
 }
 
+export async function packagesNeedInstalled(directory) {
+  const exec = util.promisify(execAsync);
```
That is how I originally had it functioning, but it makes the testing near impossible 😞, so I went for the more testable approach.

joe-yeager commented:
> I deleted my hubspot.config.yml file and ran the command. It correctly threw an error, but I think we could give a more useful message here. This is what the output looks like: <img alt="image" width="375" src="https://private-user-images.githubusercontent.com/6654014/384443649-662ec19c-6838-42dd-943f-7ae165e04abb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwODQ3MTAsIm5iZiI6MTczMTA4NDQxMCwicGF0aCI6Ii82NjU0MDE0LzM4NDQ0MzY0OS02NjJlYzE5Yy02ODM4LTQyZGQtOTQzZi03YWUxNjVlMDRhYmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDhUMTY0NjUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmUwZmZmNjU0NTI2MGVkZmEzNDUyMjhiZTA5MzdjMzhhYWQyOGNmZDlkYjViMmU1YjgxYjFmN2Q5NzlkMWZmNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.mX_w4mhMwYJ-4v0mt6qVrmClwKaOYNczu-zwZ1UJFPU"> It might be nice to make sure that the hubspot.config.yml file exists before continuing to the access token checks

Good catch @brandenrodgers!

I added a check to make sure the config file exists and if it doesn't log a message.  This content is just a place holder, I'll run the content by Sandra
![Screenshot 2024-11-08 at 10 49 45 AM](https://github.com/user-attachments/assets/04b5d637-1087-4d29-92e4-620dd1c9b31a)

camden11 commented on lib/doctor/DiagnosticInfoBuilder.ts:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  async generateDiagnosticInfo(): Promise<DiagnosticInfo> {
+    this._projectConfig = await getProjectConfig(null);
```
nit: i think it would be a little clearer to just do `getProjectConfig()`. this is done a few other times in the repo

joe-yeager replied:
I'm getting a TS error if I omit the arg since it's not optional.  I can just add a `// @ts-expect-error` so when this method is type that can be deleted

camden11 replied:
Ohh weird, I didn't realized TS had any opinions about stuff defined in `// @ts-nocheck` files. Fine by me either way then

camden11 commented on lib/doctor/DiagnosticInfoBuilder.ts:
```diff
@@ -0,0 +1,214 @@
+
+export type ProjectConfig = Awaited<ReturnType<typeof getProjectConfig>>;
+
+const hubspotCli = '@hubspot/cli';
```
I can't remember if I commented this last time, so ignore me if I did and you had a good reason not to. But could we import this from package.json?

joe-yeager replied:
Yep, I'll update it!  I'm using the pakage.name elsewhere, so not sure how this one hung around

joe-yeager replied:
Oh, now I remember why.  This value is used in a type as a key, so I need to hardcode it in at least one place, but I'll update the rest

camden11 reviewed (approved):
I have a few minor comments, but nothing blocking. Overall this looks great! Super exciting to finally be getting this out soon.

joe-yeager commented on lib/doctor/Doctor.ts:
```diff
@@ -0,0 +1,385 @@
+    this.diagnosticInfoBuilder = diagnosticInfoBuilder;
+  }
+
+  async diagnose() {
```
Whoops, I did miss this.  I'm fine with `Promise<DiagnosticInfo>` 👍
